drivers/place/rooms/present.cr

```diff
@@ -0,0 +1,157 @@
+require "placeos-driver"
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/switchable"
+
+struct Enum
+  def to_json(json : JSON::Builder)
+    json.string(to_s.underscore)
+  end
+end
+
+class Place::Rooms::Present < PlaceOS::Driver
+  generic_name :System
+  descriptive_name "Presentation space logic"
+  description <<-DESC
+    Room level state and behaviours.
+
+    This driver provides a high-level API for interaction with devices, systems
+    and integrations found within common presentation spaces. It's behavior will
+    adapt to match the capabilities and configuration of other drivers present
+    in the same system.
+    DESC
+
+  accessor proj : Projector_1, implementing: InputSelection
+  accessor prev : Display_1, implementing: InputSelection
+  accessor lcd1 : Display_2, implementing: InputSelection
+  accessor lcd2 : Display_3, implementing: InputSelection
+  accessor lcd3 : Display_4, implementing: InputSelection
+  accessor switcher : Switcher_1, implementing: Switchable
+
+  def on_load
+    load_mock_state
+    on_update
+  end
+
+  enum InputType
+    Laptop
+    Wireless
+    PC
+    DocCam
+  end
+
+  enum OutputType
+    LCD
+    Projector
+  end
+
+  abstract class Node(Type)
+    include JSON::Serializable
+    property name : String
+    property type : Type
+    property mute : Bool? = nil
+    property volume : Int32? = nil
+    property n_id : UInt64 = rand(UInt64)
+  end
+
+  class Input < Node(InputType)
+    property outputs = [] of String
+    def initialize(@name, @type, @outputs); end
+  end
+
+  class Output < Node(OutputType)
+    property mod : String
+    property source : String? = nil
+    property inputs = [] of String
+    def initialize(@name, @type, @mod, @inputs); end
+  end
+
+  @inputs = {
+    "vga" => Input.new("VGA", InputType::Laptop, ["proj", "prev"]),
+    "hdmi" => Input.new("HDMI", InputType::Laptop, ["proj", "prev"]),
+    "pc" => Input.new("PC", InputType::PC, ["proj", "prev"]),
+    "doccam" => Input.new("Visualiser", InputType::DocCam, ["proj", "prev"]),
+    "opc1" => Input.new("PC", InputType::PC, ["lcd1"]),
+    "opc2" => Input.new("PC", InputType::PC, ["lcd2"]),
+    "opc3" => Input.new("PC", InputType::PC, ["lcd3"])
+  }
+
+  @outputs = {
+    "proj" => Output.new("Projector", OutputType::LCD, "Projector_1", ["vga", "hdmi", "doccam"]),
+    "prev" => Output.new("LCD", OutputType::LCD, "Display_1", ["vga", "hdmi", "doccam"]),
+    "lcd1" => Output.new("LCD", OutputType::LCD, "Display_2", ["opc1"]),
+    "lcd2" => Output.new("LCD", OutputType::LCD, "Display_3", ["opc2"]),
+    "lcd3" => Output.new("LCD", OutputType::LCD, "Display_4", ["opc3"])
+  }
```

Comment:
i don't see this changing, any reason a NamedTuple can't be used here and above?